File: jobs/gmdh_fraud_trainer.py
"""
GMDH Fraud Trainer (Model A) — Python implementation.

2-layer self-organizing polynomial model for fraud detection.
Inputs: semantic_risk (from Bedrock), velocity_1h, proxy_score, amount_deviation
Output: fraud probability (0-1)

Same algorithm as Scala version but runs natively in Airflow container.
"""
import json
import numpy as np
import os


def train_fraud_model(data_path=None, output_path=None):
    """Train 2-layer GMDH model on fraud dataset."""

    data_path = data_path or '/opt/airflow/project/data/fraud_transactions.csv'
    output_path = output_path or '/opt/airflow/project/data/fraud_model_coeffs.json'

    # Load CSV
    import csv
    with open(data_path, 'r') as f:
        reader = csv.DictReader(f)
        rows = list(reader)

    n = len(rows)
    x1 = np.array([float(r['semantic_risk']) for r in rows])
    x2 = np.array([float(r['velocity_1h']) / 50.0 for r in rows])
    x3 = np.array([float(r['proxy_score']) for r in rows])
    x4 = np.array([float(r['amount_deviation']) for r in rows])
    y = np.array([float(r['is_fraud']) for r in rows])

    # Train/validation split (70/30)
    split = int(n * 0.7)
    X_all = np.column_stack([x1, x2, x3, x4])
    X_train, X_val = X_all[:split], X_all[split:]
    y_train, y_val = y[:split], y[split:]

    col_names = ['x1', 'x2', 'x3', 'x4']

    def fit_node(X_tr, y_tr, X_vl, y_vl, i, j):
        """Fit one GMDH node: y = b0 + b1*xi + b2*xj + b3*xi*xj"""
        xi_tr, xj_tr = X_tr[:, i], X_tr[:, j]
        xi_vl, xj_vl = X_vl[:, i], X_vl[:, j]

        # Design matrix: [1, xi, xj, xi*xj]
        A_tr = np.column_stack([np.ones(len(xi_tr)), xi_tr, xj_tr, xi_tr * xj_tr])
        A_vl = np.column_stack([np.ones(len(xi_vl)), xi_vl, xj_vl, xi_vl * xj_vl])

        # Least squares
        coeffs, _, _, _ = np.linalg.lstsq(A_tr, y_tr, rcond=None)

        # RMSE on validation
        pred = A_vl @ coeffs
        rmse = np.sqrt(np.mean((pred - y_vl) ** 2))

        return {
            'name': f'node_{col_names[i]}_{col_names[j]}',
            'rmse': rmse,
            'intercept': coeffs[0],
            'coeffs': coeffs[1:].tolist(),
            'inputs': [col_names[i], col_names[j]],
            'i': i, 'j': j
        }

    # Layer 1: All C(4,2) = 6 pairwise nodes
    nodes = []
    for i in range(4):
        for j in range(i + 1, 4):
            node = fit_node(X_train, y_train, X_val, y_val, i, j)
            nodes.append(node)

    # Selection: top 2 by lowest RMSE
    nodes.sort(key=lambda n: n['rmse'])
    w1, w2 = nodes[0], nodes[1]

    print(f"Layer 1 winners: {w1['name']} (RMSE={w1['rmse']:.4f}), {w2['name']} (RMSE={w2['rmse']:.4f})")

    # Compute z1, z2 for Layer 2
    def compute_z(X, node):
        xi, xj = X[:, node['i']], X[:, node['j']]
        return node['intercept'] + node['coeffs'][0] * xi + node['coeffs'][1] * xj + node['coeffs'][2] * xi * xj

    z1_train = compute_z(X_train, w1)
    z2_train = compute_z(X_train, w2)

    # Layer 2: Master node — y = b0 + b1*z1 + b2*z2 + b3*z1*z2
    A_master = np.column_stack([np.ones(len(z1_train)), z1_train, z2_train, z1_train * z2_train])
    master_coeffs, _, _, _ = np.linalg.lstsq(A_master, y_train, rcond=None)

    # Validate final model
    z1_val = compute_z(X_val, w1)
    z2_val = compute_z(X_val, w2)
    A_val = np.column_stack([np.ones(len(z1_val)), z1_val, z2_val, z1_val * z2_val])
    final_pred = A_val @ master_coeffs
    final_rmse = np.sqrt(np.mean((final_pred - y_val) ** 2))

    print(f"Layer 2 master node RMSE: {final_rmse:.4f}")

    # Export model
    model = {
        'beta0': float(master_coeffs[0]),
        'betas': [float(c) for c in master_coeffs[1:]],
        'layer1': [
            {'node': w1['name'], 'intercept': float(w1['intercept']),
             'coeffs': w1['coeffs'], 'inputs': w1['inputs']},
            {'node': w2['name'], 'intercept': float(w2['intercept']),
             'coeffs': w2['coeffs'], 'inputs': w2['inputs']}
        ],
        'final_rmse': float(final_rmse),
        'training_samples': split
    }

    os.makedirs(os.path.dirname(output_path), exist_ok=True)
    with open(output_path, 'w') as f:
        json.dump(model, f, indent=2)

    print(f"Fraud Model A saved to: {output_path}")
    return model


if __name__ == '__main__':
    train_fraud_model(
        data_path='data/fraud_transactions.csv',
        output_path='data/fraud_model_coeffs.json'
    )

File: jobs/gmdh_fraud_trainer.scala
import org.apache.spark.sql.{DataFrame, SparkSession}
import org.apache.spark.sql.functions._
import org.apache.spark.ml.regression.LinearRegression
import org.apache.spark.ml.feature.VectorAssembler
import org.json4s.jackson.Serialization
import org.json4s.jackson.Serialization.writePretty
import org.json4s.NoTypeHints
import java.io.PrintWriter
import java.io.File

/**
 * GMDH Fraud Trainer (Model A)
 *
 * 2-layer self-organizing polynomial model for fraud detection.
 * Inputs: semantic_risk (from Bedrock), velocity_1h, proxy_score, amount_deviation
 * Output: fraud probability (0-1)
 *
 * Architecture mirrors Model B (System Efficiency) but trained on fraud features.
 */
object FraudTrainer {
  implicit val formats = Serialization.formats(NoTypeHints)

  def main(args: Array[String]): Unit = {
    val spark = SparkSession.builder
      .appName("GMDH_Fraud_Trainer")
      .master("local[*]")
      .getOrCreate()

    try {
      val csvPath = if (args.nonEmpty) args(0) else "data/fraud_transactions.csv"
      val outputPath = if (args.length > 1) args(1) else "data/fraud_model_coeffs.json"

      val df = spark.read.option("header", "true").option("inferSchema", "true").csv(csvPath)

      val ds = df.select(
        col("semantic_risk").as("x1"),
        (col("velocity_1h") / 50.0).as("x2"),
        col("proxy_score").as("x3"),
        col("amount_deviation").as("x4"),
        col("is_fraud").as("label")
      ).na.fill(0.0)

      val Array(train, valid) = ds.randomSplit(Array(0.7, 0.3), seed = 42)
      val colNames = Array("x1", "x2", "x3", "x4")

      // Layer 1: All C(4,2) = 6 pairwise nodes
      val nodes = for {
        i <- 0 until colNames.length
        j <- (i + 1) until colNames.length
      } yield {
        val cI = colNames(i)
        val cJ = colNames(j)
        val dft = train.withColumn("interaction", col(cI) * col(cJ))
        val ass = new VectorAssembler().setInputCols(Array(cI, cJ, "interaction")).setOutputCol("f")
        val model = new LinearRegression().setLabelCol("label").setFeaturesCol("f").fit(ass.transform(dft))
        val vt = valid.withColumn("interaction", col(cI) * col(cJ))
        val rmse = model.evaluate(ass.transform(vt)).rootMeanSquaredError
        (s"node_${cI}_${cJ}", rmse, model.intercept, model.coefficients.toArray, cI, cJ)
      }

      // Selection: top 2 by lowest RMSE
      val winners = nodes.sortBy(_._2).take(2)
      val w1 = winners(0)
      val w2 = winners(1)

      println(s"Layer 1 winners: ${w1._1} (RMSE=${w1._2}), ${w2._1} (RMSE=${w2._2})")

      // Layer 2: Master node combining z1, z2
      def calc(df: DataFrame, w: (String, Double, Double, Array[Double], String, String), out: String): DataFrame = {
        df.withColumn(out, lit(w._3) + col(w._5) * w._4(0) + col(w._6) * w._4(1) + col(w._5) * col(w._6) * w._4(2))
      }

      val l2 = calc(calc(train, w1, "z1"), w2, "z2").withColumn("interaction", col("z1") * col("z2"))
      val masterModel = new LinearRegression().setLabelCol("label").setFeaturesCol("f").fit(
        new VectorAssembler().setInputCols(Array("z1", "z2", "interaction")).setOutputCol("f").transform(l2)
      )

      // Export as flat JSON for fast inference
      val export = Map(
        "beta0" -> masterModel.intercept,
        "betas" -> masterModel.coefficients.toArray.toList,
        "layer1" -> List(
          Map("node" -> w1._1, "intercept" -> w1._3, "coeffs" -> w1._4.toList, "inputs" -> List(w1._5, w1._6)),
          Map("node" -> w2._1, "intercept" -> w2._3, "coeffs" -> w2._4.toList, "inputs" -> List(w2._5, w2._6))
        ),
        "timestamp" -> System.currentTimeMillis()
      )

      val pw = new PrintWriter(new File(outputPath))
      pw.write(writePretty(export))
      pw.close()

      println(s"✅ Fraud Model A saved to: $outputPath")

    } finally {
      spark.stop()
    }
  }
}

FraudTrainer.main(Array("data/fraud_transactions.csv", "data/fraud_model_coeffs.json"))
